### ISSUE TITLE
Testing internationalized characters is broken with PostgreSQL

### ISSUE THREAD

**Issue by pnax** [zonemaster/zonemaster-backend#931]
Testing a domain such as `café.fr` breaks the Backend.
https://zonemaster.net/result/bb711a47bf7f203f

The `params` value is stored differently in database, as string in SQLite and MySQL, as a JSON object in PostgreSQL. This has an impact on how each driver retrieves the data from database in `get_test_params`.

## Playing with the drivers

It is possible to inspect the value of `$params_json` returned by each driver with the module [Data::Peek](https://metacpan.org/pod/Data::Peek)

https://github.com/zonemaster/zonemaster-backend/blob/b8df4d0ad702cbe05e0dddea3530f3c7ffd8ba68/lib/Zonemaster/Backend/DB.pm#L256

The following patch is applied to the code, to dive into the `$params_json` string:

```diff
diff --git a/lib/Zonemaster/Backend/DB.pm b/lib/Zonemaster/Backend/DB.pm
index 3898ab8..df4975c 100644
--- a/lib/Zonemaster/Backend/DB.pm
+++ b/lib/Zonemaster/Backend/DB.pm
@@ -14,6 +14,8 @@ use Log::Any qw( $log );
 use Zonemaster::Engine::Profile;
 use Zonemaster::Backend::Errors;
 
+use Data::Peek;
+
 requires qw(
   add_batch_job
   create_db
@@ -258,6 +260,8 @@ sub get_test_params {
     die Zonemaster::Backend::Error::ResourceNotFound->new( message => "Test not found", data => { test_id => $test_id } )
         unless defined $params_json;
 
+    print DPeek( $params_json ) . "\n";
+
     my $result;
     eval {
         # TODO: do we use "encode_utf8" as this was the case in PostgreSQL
```

Then foreach database, testing with the `café.fr` domain will give the outputs listed below

```sh
curl -sH "Content-Type: application/json" --data '{"method":"start_domain_test","params":{"domain":"café.fr"},"id":1,"jsonrpc":"2.0"}' localhost:5000/api
```

| engine | output |
| --- | --- |
| SQLite | `PV("{\"domain\":\"caf\303\251.fr\",\"ds_info\":[],\"ipv4\":true,\"ipv6\":true,\"nameservers\":[],\"priority\":10,\"profile\":\"defa"...\0)`
| MySQL | `PV("{\"domain\":\"caf\303\251.fr\",\"ds_info\":[],\"ipv4\":true,\"ipv6\":true,\"nameservers\":[],\"priority\":10,\"profile\":\"defa"...\0)` |
| PostgreSQL | `PV("{\"domain\":\"caf\303\251.fr\",\"ds_info\":[],\"ipv4\":true,\"ipv6\":true,\"nameservers\":[],\"priority\":10,\"profile\":\"defa"...\0) [UTF8 "{"domain":"caf\x{e9}.fr","ds_info":[],"ipv4":true,"ipv6":true,"nameservers":[],"priority":10,"profile":"default","queue":0}"]` |

The Perl's internal representation of the string is different. The string retrieved with DBD::pg has an UTF-8 section in the DPeek output that is not present with SQLite and MySQL. This requires to be careful when calling [`decode_json`](https://metacpan.org/pod/JSON::PP#decode_json) that _"expects an UTF-8 (binary) string"_

**Comment by matsduf** [comment]
This i unfortunately not the first time this happens. zonemaster-cli can handle it. Here and now we have to solve the issue, but to get a stable solution we should merge and implement the proposal in https://github.com/zonemaster/zonemaster/pull/942

**Comment by matsduf** [comment]
Resolved by #932.
